src/python/map/rings.py

```diff
@@ -0,0 +1,160 @@
+    rank, n_mpi_processes, ints
+        rank of the current MPI process and total number of processes
+
+    Returns
+    -------
+
+    grid : libsharp.healpix_grid
+        libsharp object that includes metadata about HEALPix distributed rings
+    local_pix : np.ndarray
+        integer array of local pixel indices in the current MPI process in RING
+        ordering
+    """
+    nrings = 4 * nside - 1  # four missing pixels
+
+    # ring indices are 1-based
+    ring_indices_emisphere = np.arange(2*nside, dtype=np.int32) + 1
+
+    local_ring_indices = ring_indices_emisphere[rank::n_mpi_processes]
+
+    # to improve performance, simmetric rings north/south need to be in the same rank
+    # therefore we use symmetry to create the full ring indexing
+
+    if local_ring_indices[-1] == 2 * nside:
+        # has equator ring
+        local_ring_indices = np.concatenate(
+          [local_ring_indices[:-1],
+           nrings - local_ring_indices[::-1] + 1]
+        )
+    else:
+        # does not have equator ring
+        local_ring_indices = np.concatenate(
+          [local_ring_indices,
+           nrings - local_ring_indices[::-1] + 1]
+        )
+
+    grid = libsharp.healpix_grid(nside, rings=local_ring_indices)
+    return grid, local_ring_indices
+
+class DistRings(object):
+    """
+    A map with unique pixels distributed as disjoint isolatitude rings.
+
+    Designed for Harmonic Transforms with libsharp
+
+    Pixel domain data is distributed across an MPI communicator.  Each
+    process has a number of isolatitude rings and a list of the pixels
+    within those rings.
+
+    Args:
+        comm (mpi4py.MPI.Comm): the MPI communicator containing all
+            processes.
+        size (int): the total number of pixels.
+        nnz (int): the number of values per pixel.
+        submap (int): the locally stored data is in units of this size.
+        local (array): the list of local submaps (integers).
+        localpix (array): the list of local pixels (integers).
+        nest (bool): nested pixel order flag
+    """
+    def __init__(self, comm=MPI.COMM_WORLD, nnz=1, dtype=np.float64,
+                 nside=16):
+        self.data = None
+        self._comm = comm
+        self._nnz = nnz
+        self._dtype = dtype
+        self._nest = False
+        self._nside = nside
+
+        self._cache = Cache()
+
+        self._libsharp_grid, self._local_ring_indices = distribute_rings(
+                self._nside, self._comm.rank, self._comm.size)
+        # returns start index of the ring and number of pixels
+        startpix, ringpix, _, _, _ = hp.ringinfo(
+                self._nside, self._local_ring_indices.astype(np.int64))
+
+        local_npix = self._libsharp_grid.local_size()
+        self._local_pixels = self._cache.create(
+                "local_pixels", shape=(local_npix,), type=np.int64)
+        expand_pix(startpix, ringpix, local_npix, self._local_pixels)
+
+        self.data = self._cache.create(
```

Comment:
@tskisner quick question: how should I store maps for each channel? should I get `num_channels` as input parameter to `DistRings` and then create a 3D `data` buffer `(npix, nnz, num_channels)`?